### PR TITLE
Blobstore library implementation.

### PR DESCRIPTION
**Pull request by sofifch**
I just implemented the feature of uploading images along with comments in my page. However, I still have not managed to display the image as an html element, but I do know that I am getting its url correctly by checking its value through the data admin.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    BlobstoreService myBlobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    String myUploadUrl = myBlobstoreService.createUploadUrl("/data");
+
+    response.setContentType("text/html");
+    response.getWriter().println(myUploadUrl);
+  }
+}
```
Comment (rjkroege) [inline portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java]: missing newline

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -60,13 +68,52 @@ private static String convertToJsonWithGson(List<String> arraylist) {
...
       response.sendRedirect("/comments.html");
   }
+
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get("image");
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
```
Comment (rjkroege) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:98]: nit: submitted the form

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -60,13 +68,52 @@ private static String convertToJsonWithGson(List<String> arraylist) {
...
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // We could check the validity of the file here, e.g. to make sure it's an image file
+    // https://stackoverflow.com/q/10779564/873165
```
Comment (rjkroege) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:114]: This seems difficult. Google source conventionally notes this kind of thing with TODO(person who might fix) and have an issue number for it.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -60,13 +68,52 @@ private static String convertToJsonWithGson(List<String> arraylist) {
...
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+    String url = imagesService.getServingUrl(options);
+
+    // GCS's localhost preview is not actually on localhost,
+    // so make the URL relative to the current domain.
+    if(url.startsWith("http://localhost:8080/")){
```
Comment (rjkroege) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:123]: What is happening here?

**Comment by rjkroege** [comment]
Per chat: perhaps the image needs a mime type set?

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java (hunk 1 above)
Comment (sofifch) [reply portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java]: Added!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (sofifch) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:98]: Oh I got this from an example that was part of the tutorial itself, but I will add the "the" :)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (sofifch) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:114]: Yes, I think that's why they restrained from including the implementation for that.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (sofifch) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:123]: I am not very sure, I took it from the example on the tutorial. I can make some research to understand this 3 lines better.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -60,13 +75,54 @@ private static String convertToJsonWithGson(List<String> arraylist) {
...
+      return null;
+    }
+
+    // We could check the validity of the file here, e.g. to make sure it's an image file
+    // https://stackoverflow.com/q/10779564/873165
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+    String url = imagesService.getServingUrl(options);
```
Comment (rjkroege) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:119]: Ah. This makes sense.

**Review by rjkroege — APPROVED** [review]

**Pull request merged**